File: app/src/main/java/ru/mail/sporttogether/auth/google/GoogleSocialNetwork.java
package ru.mail.sporttogether.auth.google;

import android.app.Activity;
import android.content.Intent;
import android.content.IntentSender;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.util.Log;

import com.google.android.gms.auth.api.Auth;
import com.google.android.gms.auth.api.signin.GoogleSignInOptions;
import com.google.android.gms.auth.api.signin.GoogleSignInResult;
import com.google.android.gms.common.ConnectionResult;
import com.google.android.gms.common.api.GoogleApiClient;
import com.google.android.gms.common.api.ResultCallback;
import com.google.android.gms.common.api.Scope;
import com.google.android.gms.common.api.Status;

import ru.mail.sporttogether.auth.core.AuthError;
import ru.mail.sporttogether.auth.core.SocialNetwork;
import ru.mail.sporttogether.auth.core.SocialNetworkManager;

/**
 * Created by said on 07.11.16.
 */

public class GoogleSocialNetwork implements SocialNetwork, GoogleApiClient.ConnectionCallbacks, GoogleApiClient.OnConnectionFailedListener {

    private static final String TAG = GoogleSocialNetwork.class.getSimpleName();
    static final int REQUEST_RESOLVE_ERROR = 1001;
    public static final int ID = 2;
    private Scope[] scopes;
    private GoogleApiClient googleApiClient;
    private String appID;
    private Activity activity;
    public final static int RC_SIGN_IN = 201;
    private boolean resolvingError;
    private LoginCallback callback;

    public GoogleSocialNetwork(Activity activity, String appID, Scope[] scopes, LoginCallback callback) {
        this.scopes = scopes;
        this.appID = appID;
        this.callback = callback;
        this.activity = activity;
        googleApiClient = createGoogleAPIClient(activity);
    }

    @Override
    public boolean isLoginOK(int requestCode, int resultCode, Intent data) {
        if (requestCode == REQUEST_RESOLVE_ERROR) {
            resolvingError = false;
            if (resultCode == Activity.RESULT_OK) {
                connectAndRequestGoogleAccount(RC_SIGN_IN, REQUEST_RESOLVE_ERROR);
            }
            return true;
        } else if (requestCode == RC_SIGN_IN) {
            if (resultCode == Activity.RESULT_OK) {
                final GoogleSignInResult result = Auth.GoogleSignInApi.getSignInResultFromIntent(data);
                if (result.isSuccess()) {
                    callback.onOK();
                } else if (result.getStatus().isCanceled()) {
                    callback.onCancel();
                }
            }
            return true;
        }

        return false;
    }

    private GoogleApiClient createGoogleAPIClient(Activity activity) {
        final GoogleSignInOptions.Builder gsoBuilder = new GoogleSignInOptions.Builder()
                .requestIdToken(appID);
        if (scopes.length == 1) {
            gsoBuilder.requestScopes(scopes[0]);
        } else if (scopes.length > 1) {
            gsoBuilder.requestScopes(scopes[0], scopes);
        }

        final GoogleApiClient.Builder builder = new GoogleApiClient.Builder(activity, this, this)
                .addApi(Auth.GOOGLE_SIGN_IN_API, gsoBuilder.build());

        return builder.build();
    }

    @Override
    public Integer getID() {
        return ID;
    }

    @Override
    public void login(Activity activity) {
        SocialNetworkManager.getInstance().setSocialNetworkID(ID);
        connectAndRequestGoogleAccount(RC_SIGN_IN, REQUEST_RESOLVE_ERROR);
    }

    @Override
    public boolean isLoggedIn() {
        String token = SocialNetworkManager.getInstance().getSharedPreferces().getString(SocialNetworkManager.AUTH_TOKEN, null);
        return token != null;
    }

    @Override
    public void logout() {
        try {
            Auth.GoogleSignInApi.signOut(googleApiClient).setResultCallback(new ResultCallback<Status>() {
                @Override
                public void onResult(@NonNull Status status) {
                    if (!status.isSuccess()) {
                        Log.w(TAG, "Couldn't clear account and credentials");
                    } else {
                        SocialNetworkManager.getInstance().getSharedPreferces()
                                .edit()
                                .putInt(SocialNetworkManager.AUTH_ID, 0)
                                .putString(SocialNetworkManager.AUTH_TOKEN, null)
                                .apply();
                    }
                }
            });
        } catch (IllegalStateException e) {
            Log.e(TAG, "Failed to clear the Google Plus Session", e);
        }
    }

    @Override
    public void sharePost(Activity activity) {

    }

    private void connectAndRequestGoogleAccount(int signInRequestCode, int errorResolutionRequestCode) {
        if (googleApiClient.isConnected()) {
            requestGoogleAccount(signInRequestCode);
        } else if (!googleApiClient.isConnecting()) {
            googleApiClient.connect();
        }
    }

    private void requestGoogleAccount(int signInRequestCode) {
        final Intent signInIntent = Auth.GoogleSignInApi.getSignInIntent(googleApiClient);
        activity.startActivityForResult(signInIntent, signInRequestCode);
    }

    @Override
    public void onConnected(@Nullable Bundle bundle) {
        requestGoogleAccount(RC_SIGN_IN);
    }

    @Override
    public void onConnectionSuspended(int i) {

    }

    @Override
    public void onConnectionFailed(@NonNull ConnectionResult connectionResult) {
        if (resolvingError) {
            // Already attempting to resolve an error.
            return;
        }
        if (connectionResult.hasResolution()) {
            Log.v(TAG, "Connection failed. Trying to start the resolution.");
            try {
                resolvingError = true;
                connectionResult.startResolutionForResult(activity, REQUEST_RESOLVE_ERROR);
            } catch (IntentSender.SendIntentException e) {
                googleApiClient.connect();
            }
        } else {
            Log.v(TAG, "Connection failed. No resolution was possible.");
            callback.onError(new AuthError(connectionResult.getErrorMessage(), connectionResult.getErrorCode()));
            resolvingError = true;
        }
    }
}

File: app/src/main/java/ru/mail/sporttogether/auth/core/social_networks/GoogleSocialNetwork.kt
package ru.mail.sporttogether.auth.core.social_networks

import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.util.Log
import com.crashlytics.android.answers.Answers
import com.crashlytics.android.answers.LoginEvent
import com.google.android.gms.auth.api.Auth
import com.google.android.gms.auth.api.signin.GoogleSignInAccount
import com.google.android.gms.auth.api.signin.GoogleSignInOptions
import com.google.android.gms.auth.api.signin.GoogleSignInResult
import com.google.android.gms.common.ConnectionResult
import com.google.android.gms.common.api.GoogleApiClient
import ru.mail.sporttogether.BuildConfig
import ru.mail.sporttogether.auth.core.ISocialNetwork
import ru.mail.sporttogether.auth.core.SocialNetworkError
import ru.mail.sporttogether.auth.core.SocialNetworkManager
import ru.mail.sporttogether.auth.core.SocialPerson
import ru.mail.sporttogether.auth.core.listeners.OnLoginCompleteListener
import ru.mail.sporttogether.auth.core.listeners.OnRequestSocialPersonCompleteListener


/**
 * Created by said on 17.11.16
 * это дичайший говнокод, но нет времени рефачить((
 */

class GoogleSocialNetwork(val context: Context) : ISocialNetwork, GoogleApiClient.OnConnectionFailedListener {

    private var onLoginCompleteListener: OnLoginCompleteListener? = null
    private var sharedPreferences: SharedPreferences


    private val gso = GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
            .requestEmail()
            .requestId()
            .requestProfile()
            .build()

    private val googleApiClient = GoogleApiClient.Builder(context)
            .addApi(Auth.GOOGLE_SIGN_IN_API, gso)
            .build()

    override val id: Int
        get() = ID

    override val token: String
        get() = sharedPreferences.getString(KEY_TOKEN, "")

    private var acct: GoogleSignInAccount? = null


    init {
        this.sharedPreferences = context.getSharedPreferences(SocialNetworkManager.SHARED_PREFERCE_TAG, Context.MODE_PRIVATE)

    }

    override fun tryAutoLogin(onLoginCompleteListener: OnLoginCompleteListener) =
            if (isConnected) {
                onLoginCompleteListener.onSuccess(GoogleSocialNetwork.ID)
                true
            } else false

    override fun logout() {
        googleApiClient.connect()
        googleApiClient.registerConnectionCallbacks(object : GoogleApiClient.ConnectionCallbacks {
            override fun onConnected(p0: Bundle?) {
                Auth.GoogleSignInApi.signOut(googleApiClient).setResultCallback { status ->

                }
            }

            override fun onConnectionSuspended(p0: Int) {

            }
        })

    }

    override fun login(activity: AppCompatActivity, onLoginCompleteListener: OnLoginCompleteListener) {
        this.onLoginCompleteListener = onLoginCompleteListener
        val signInIntent = Auth.GoogleSignInApi.getSignInIntent(googleApiClient)
        activity.startActivityForResult(signInIntent, RC_SIGN_IN)
    }

    override fun onConnectionFailed(p0: ConnectionResult) {

    }

    override fun onDestroy() {

    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode == RC_SIGN_IN) {
            val result = Auth.GoogleSignInApi.getSignInResultFromIntent(data)
            handleSignInResult(result)
            return
        }
    }

    private fun handleSignInResult(result: GoogleSignInResult) {
        Log.d(TAG, "handleSignInResult:" + result.isSuccess)
        if (result.isSuccess) {
            acct = result.signInAccount
            putToSharedPreferences(acct)
            onLoginCompleteListener!!.onSuccess(ID)
            if (!BuildConfig.DEBUG) Answers.getInstance().logLogin(LoginEvent().putMethod("Google").putSuccess(true))
        } else {
            onLoginCompleteListener!!.onError(SocialNetworkError(result.status.statusMessage, result.status.statusCode))
        }
    }

    private fun putToSharedPreferences(account: GoogleSignInAccount?) {
        account?.let {
            sharedPreferences.edit()
                    .putString(KEY_ID, "g|" + it.id)
                    .putString(KEY_PHOTO, it.photoUrl.toString())
                    .putString(KEY_NAME, it.displayName)
                    .putString(KEY_TOKEN, it.email)
                    .commit()
        }
    }

    override val isConnected: Boolean
        get() = !sharedPreferences.getString(KEY_ID, "").isNullOrEmpty()


    override fun requestPerson(onRequestSocialPersonCompleteListener: OnRequestSocialPersonCompleteListener) {
        val person = SocialPerson()
        person.id = sharedPreferences.getString(KEY_ID, "")
        if (!person.id.isNullOrEmpty()) {
            person.avatarURL = sharedPreferences.getString(KEY_PHOTO, "")
            person.name = sharedPreferences.getString(KEY_NAME, "")
            person.email = sharedPreferences.getString(KEY_TOKEN, "")
            onRequestSocialPersonCompleteListener.onComplete(person, ID)
            return
        }
        onRequestSocialPersonCompleteListener.onError(SocialNetworkError("Not exists", -1))
    }

    companion object {
        @JvmStatic private val RC_SIGN_IN = 1824
        @JvmStatic val ID = 3
        @JvmStatic private val TAG = "GoogleSocialNetwork"

        @JvmStatic private val KEY_ID = "KEY_ID"
        @JvmStatic private val KEY_PHOTO = "KEY_PHOTO"
        @JvmStatic private val KEY_NAME = "KEY_NAME"
        @JvmStatic private val KEY_TOKEN = "KEY_TOKEN"
    }
}
